lib/core/network/protocols/http+websocket.js

```diff
@@ -0,0 +1,920 @@
+
+    const socket = this.socketByConnectionId.get(connectionId);
+
+    if (!socket) {
+      return;
+    }
+
+    socket.end(1011, message ? Buffer.from(message) : WS_GENERIC_CLOSE_MESSAGE);
+  }
+
+  wsOnOpenHandler (socket) {
+    const ip = Buffer.from(socket.getRemoteAddressAsText()).toString();
+    const connection = new ClientConnection(this.name, [ip]);
+
+    this.entryPoint.newConnection(connection);
+    this.connectionBySocket.set(socket, connection);
+    this.socketByConnectionId.set(connection.id, socket);
+    this.backpressureBuffer.set(socket, []);
+  }
+
+  wsOnCloseHandler (socket, code, message) {
+    const connection = this.connectionBySocket.get(socket);
+
+    if (!connection) {
+      return;
+    }
+
+    if (debugWS.enabled) {
+      debugWS(
+        '[%s] received a `close` event (CODE: %d, REASON: %s)',
+        connection.id,
+        code,
+        Buffer.from(message || '').toString());
+    }
+    this.entryPoint.removeConnection(connection.id);
+    this.connectionBySocket.delete(socket);
+    this.backpressureBuffer.delete(socket);
+    this.socketByConnectionId.delete(connection.id);
+  }
+
+  wsOnMessageHandler (socket, data) {
+    const connection = this.connectionBySocket.get(socket);
+
+    if (!data || data.byteLength === 0 || !connection) {
+      return;
+    }
+
+    // enforce rate limits
+    if (this.wsConfig.rateLimit > 0) {
+      if (socket.last === this.now) {
+        socket.count++;
+
+        if (socket.count > this.wsConfig.rateLimit) {
+          this.wsSendError(socket, connection, WS_RATE_LIMIT_EXCEEDED_ERROR);
+          return;
+        }
+      }
+      else {
+        socket.last = this.now;
+        socket.count = 1;
+      }
+    }
+
+    let parsed;
+    const message = Buffer.from(data).toString();
+
+    debugWS('[%s] client message: %s', connection.id, message);
+
+    try {
+      parsed = JSON.parse(message);
+    }
+    catch (e) {
+      /*
+       we cannot add a "room" information since we need to extract
+       a request ID from the incoming data, which is apparently
+       not a valid JSON
+       So... the error is forwarded to the client, hoping they know
+       what to do with it.
+       */
+      this.wsSendError(
+        socket,
+        connection,
+        kerrorWS.getFrom(e, 'unexpected_error', e.message));
+      return;
+    }
+
+    if (parsed.p && parsed.p === 1 && Object.keys(parsed).length === 1) {
+      debugWS('[%s] sending back a "pong" message', connection.id);
+      this.wsSend(socket, WS_APP_PONG_RESPONSE);
+      return;
+    }
+
+    try {
+      this.entryPoint.execute(new Request(parsed, { connection }), result => {
+        if (result.content && typeof result.content === 'object') {
+          result.content.room = result.requestId;
+        }
+        this.wsSend(socket, Buffer.from(JSON.stringify(result.content)));
+      });
+    }
+    catch (e) {
+      const errobj = {
+        error: {
+          message: e.message
```

Comment:
Don't we send our other error properties? Like `id`, `code`, `stack` etc?

Also, why not using the `wsSendError` method?

Reply:
Hm... This bit of code comes from our holder websocket message handler, I didn't take a good look at it. After some analysis, I don't see the point of this try..catch at all: the entire process is safe, and the callback is always invoked with an error result if something wrong happens.

I think I'll remove this bit of code altogether.

Reply:
Ah, yes, it IS useful: it's there to catch errors when promoting the payload to a KuzzleRequest object. I've properly separated the code to make it clearer, and the code now uses wsSendError, as it should have since the beginning.